## INFO284 - Group exam

Your tasks are:

#### Preprocessing: 
Get to understand the data and remove columns and rows that you do not find useful for your machine learning models. It is possible to for example focus on one type of gear, one or a few species, group categories of species, etc. etc. This will help to reduce
the data set you are working on. Understanding data and preparing them for model building
is a main task of machine learning.

#### Supervised learning: 
Build at least three machine learning models to predict or classify catch
data related to a fishery operation, i.e. species caught, amounts of one fish species (sums of amounts is also possible), or other features related to fish catches. One of the models needs to be a deep learning model.

#### Unsupervised learning: 
Build a clustering model for the data set. You may use a different preprocessing for this task than for the data in task b).


You shall deliver code in the form of a well commented Jupyter notebook. This code needs to run on the original data set, so any preprocessing you choose to do needs to be programmed in Python and included in the notebook. The code shall in the end return the results of your experiments with your chosen models. You need to explain

- Important and relevant properties of the data
- how you preprocessed data like which features you selected, did you do dimension reduction 
- how you reformatted data, etc.
- how you decided on parameters for your machine learning models,
- if you used any regularization techniques? In case how.
- how the methods were measured and compared

Please inform about any special Python libraries that need to be installed to make your code runnable.

Finally, as a concluding comment in the Jupyter notebook, you need to write a summary of your results, and discuss consequences of such results.

It is not necessarily so that high scores for machine learning models will give a good grade on your report, or vice versa, low scores a bad grade. What counts is a well-argued, well described, and smart machine learning investigation from start to end. The problem may in fact be of such nature that it is not possible to get really good results on these data.
Final note: These data are prepared for this course and are shared with you in confidence that you do not share them in any way but use them only for the purpose of this exam.


## Importerer nødvendige biblioteker
Leser csv.

In [247]:
import pandas as pd
import numpy as np
from sklearn import linear_model
from sklearn.model_selection import train_test_split
import datetime as dt
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeRegressor

df = pd.read_csv("GroupExam2024_data_and_documentation/elektronisk-rapportering-ers-2018-fangstmelding-dca-simple.csv", sep=';', decimal=',')

In [248]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 305434 entries, 0 to 305433
Data columns (total 45 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Melding ID                305434 non-null  int64  
 1   Meldingstidspunkt         305434 non-null  object 
 2   Meldingsdato              305434 non-null  object 
 3   Meldingsklokkeslett       305434 non-null  object 
 4   Starttidspunkt            305434 non-null  object 
 5   Startdato                 305434 non-null  object 
 6   Startklokkeslett          305434 non-null  object 
 7   Startposisjon bredde      305434 non-null  float64
 8   Startposisjon lengde      305434 non-null  float64
 9   Hovedområde start (kode)  303433 non-null  float64
 10  Hovedområde start         301310 non-null  object 
 11  Lokasjon start (kode)     303433 non-null  float64
 12  Havdybde start            305434 non-null  int64  
 13  Stopptidspunkt            305434 non-null  o

Av df.info ser vi de ulike kolonnene og hvilken datatype de er. og hvor mange "non-null" verdier som eksisterer. Kan se her at det eksisterer null-verdier i de fleste kolonner. Dette renses senere. Konverterer datatype etter å ha filtrert ut de uaktuelle kolonnene.

In [249]:
df.head()

Melding ID Meldingstidspunkt Meldingsdato Meldingsklokkeslett  \
0     1497177        01.01.2018   01.01.2018               00:00   
1     1497178        01.01.2018   01.01.2018               00:00   
2     1497178        01.01.2018   01.01.2018               00:00   
3     1497178        01.01.2018   01.01.2018               00:00   
4     1497178        01.01.2018   01.01.2018               00:00   

     Starttidspunkt   Startdato Startklokkeslett  Startposisjon bredde  \
0        31.12.2017  31.12.2017            00:00               -60.350   
1  30.12.2017 23:21  30.12.2017            23:21                74.885   
2  30.12.2017 23:21  30.12.2017            23:21                74.885   
3  30.12.2017 23:21  30.12.2017            23:21                74.885   
4  30.12.2017 23:21  30.12.2017            23:21                74.885   

   Startposisjon lengde  Hovedområde start (kode)  ...        Art - FDIR  \
0               -46.133                       NaN  ...  Antarktisk krill   
1                16.048                      20.0  ...              Hyse   
2                16.048                      20.0  ...             Torsk   
3                16.048                      20.0  ...         Blåkveite   
4                16.048                      20.0  ...               Sei   

   Art - gruppe (kode)      Art - gruppe  Rundvekt Lengdegruppe (kode)  \
0                506.0  Antarktisk krill  706714.0                 5.0   
1                202.0              Hyse    9594.0                 5.0   
2                201.0             Torsk    8510.0                 5.0   
3                301.0         Blåkveite     196.0                 5.0   
4                203.0               Sei     134.0                 5.0   

   Lengdegruppe  Bruttotonnasje 1969  Bruttotonnasje annen  Bredde  \
0  28 m og over               9432.0                   NaN   19.87   
1  28 m og over               1476.0                   NaN   12.60   
2  28 m og over               1476.0                   NaN   12.60   
3  28 m og over               1476.0                   NaN   12.60   
4  28 m og over               1476.0                   NaN   12.60   

   Fartøylengde  
0        133.88  
1         56.80  
2         56.80  
3         56.80  
4         56.80  

[5 rows x 45 columns]

Basert på datasettet og kolonnene vi har sett på så ønsker vi å predikere rundvekt av en aktivitet.


Ser av datasettet hvordan dataen er strukturert. 
**Organisering av datasettet**
Hver rad representerer en fangst av art. Flere rader er registrert med samme starttidspunkt eller stopptidspunkt. De ulike fangstene som er gjort på samme tidspunkt med samme melding ID kommer fra samme fiskeaktivitet. Videre så har de ulike aktivitetene som er gjort på samme Melding ID samme meldingsdato, og kommer fra samme båt.

**Bruttotonnasje 1969 og Bruttotonnasje annen**
Vi ser også at kolonnen Bruttotonnasje annen og bruttotonnasje 1969 ser ut til å utfylle hverandre.Med bruttotonnasje menes det totale volumet av hulrom i båten. Det kan ikke helt sammenlignes med lastekapasitet, men kan være en indikator. Bruttotonnasje 1969 er standarden i dag for registrering av dette målet av båten. Bruttotonnasje annen viser de båtene som er registrert gjennom andre systemer enn standarden.

Vi antar at selv om det antakelig er forskjeller i hvordan disse båtene har målt sin Bruttotonnasje, så er det sannsynligvis også så mange likheter at det ikke har betydning. Dette er det kanskje nødvendig å undersøke nermere dersom man vil ha en mer nøyaktig analyse av dataene. Disse to kolonnene vil vi derfor slå sammen. til en felles kolonne.

Hva leter vi etter?
Siden dette er en samling av fiskeridata, så tenker vi det er interessant å se nærmere på fangstdataene. Her kunne vi hatt mange interessante innfallsvinkler, men vi ønsker å se om det er mulig å predikere rundvekt av en bestemt art gitt den informasjonen vi har. Siden fiskere bruker ulike redskaper for å fange fisk kan vi kan anta at noen redskaper og metoder er mer egnet for å fange bestemte arter. For eksempel så brukes ikke teiner for å fange torsk eller trål for å fange makrell (Må sjekke opp i). I tillegg vet vi også at det foregår fiske etter ulike arter på ulike steder og tidspunkt. Skreifiske for eksempel foregår i bestemte tidsrom etter bestemte arter, med bestemte redskaper. Videre kan vi anta at det samme gjelder for for eksempel sildefiske. Kunnskap om hvordan, hvor og hva som fiskes etter er derfor viktig for å forstå datasettet. Videre avsnitt bruker vi på å undersøke informasjon omkring fisket.


In [250]:
# Først undersøker vi hvor mange registreringer det er av de ulike fisketypene og hvor mye det fiskes av hver.

species_weight_sum = df.groupby('Art - gruppe')['Rundvekt'].sum().sort_values(ascending=False)
print(df['Art - gruppe'].value_counts())
print(species_weight_sum)

Art - gruppe
Torsk                                       56574
Annen torskefisk                            45286
Sei                                         42557
Hyse                                        39120
Annen flatfisk, bunnfisk og dypvannsfisk    25267
Uer                                         19681
Steinbiter                                  16181
Dypvannsreke                                13678
Blåkveite                                    8046
Snøkrabbe                                    6070
Sild, annen                                  6006
Skater og annen bruskfisk                    4067
Makrell                                      3013
Kolmule                                      2843
Øyepål                                       2210
Annen pelagisk fisk                          1802
Vassild og strømsild                         1650
Andre skalldyr, bløtdyr og pigghuder         1329
Lodde                                        1188
Tobis og annen sil                   

Her ser vi en hovedvekt av fiske etter Torsk. På en annenplass kommer sei. Reke er det også mye aktivitet rundt.

In [251]:
# Gjør samme undersøkelse av reskaper.

redskap_weight_sum = df.groupby('Redskap FDIR')['Rundvekt'].sum().sort_values(ascending=False)
print(df['Redskap FDIR'].value_counts())
print(redskap_weight_sum)

Redskap FDIR
Bunntrål                                  120487
Snurrevad                                  51615
Andre liner                                43992
Reketrål                                   22421
Udefinert garn                             19277
Udefinert trål                             15870
Snurpenot/ringnot                          12711
Teiner                                      6818
Bunntrål par                                3335
Dobbeltrål                                  2836
Flytetrål                                   2732
Flytetrål par                               1512
Settegarn                                    649
Harpun og lignende uspesifiserte typer       494
Annet                                        462
Juksa/pilk                                    19
Dorg/harp/snik                                16
Name: count, dtype: int64
Redskap FDIR
Snurpenot/ringnot                         776961216.0
Flytetrål                                 468684685.0
Bunntrå

Her ser vi at de to mest brukte fangstredskapene er Bunntrål og Snurrevad. Videre er andre liner, reketrål og andre liner høyt oppe på listen. fra fiskeridirektoratet (kilde) vet vi at Bunntrål og Snurrevad er garn/not som blir slept langs bunnen av sjøen og samler opp fisk på den måten. Andre liner (dersom vi har forstått det riktig) er fiske med agn og krok som gjerne slepes bak båten. 

Det som er naturlig å undersøke nå er hvordan sammensetningen av fisk er med de ulike fangsmetodene, og hvor det er vanlig at de blir brukt.

In [252]:
# Ser etter hvilke fisk som blir fisket med Bunntrål:
condition = df[df['Redskap FDIR'] == 'Bunntrål']
species_weight_sum = condition.groupby('Art - gruppe')['Rundvekt'].sum().sort_values(ascending=False)
print(condition['Art - gruppe'].value_counts(), species_weight_sum)

Art - gruppe
Torsk                                       23366
Sei                                         18248
Hyse                                        16632
Annen torskefisk                            14959
Uer                                         14252
Annen flatfisk, bunnfisk og dypvannsfisk     8621
Dypvannsreke                                 7440
Steinbiter                                   7281
Blåkveite                                    5103
Skater og annen bruskfisk                     805
Øyepål                                        721
Andre skalldyr, bløtdyr og pigghuder          610
Kolmule                                       485
Vassild og strømsild                          479
Annen pelagisk fisk                           402
Sild, annen                                   320
Makrell                                       231
Tobis og annen sil                            179
Haifisk                                        64
Lodde                                

In [253]:
# Ser etter hvilke fisk som blir fisket med Snurrevad:
condition = df[df['Redskap FDIR'] == 'Snurrevad']

species_weight_sum = condition.groupby('Art - gruppe')['Rundvekt'].sum().sort_values(ascending=False)
print(condition['Art - gruppe'].value_counts(), species_weight_sum)

Art - gruppe
Torsk                                       16347
Hyse                                        11950
Sei                                         11259
Annen flatfisk, bunnfisk og dypvannsfisk     4185
Annen torskefisk                             4124
Steinbiter                                   1282
Uer                                           811
Blåkveite                                     367
Sild, annen                                    30
Makrell                                        26
Skater og annen bruskfisk                      11
Dypvannsreke                                   10
Lodde                                           9
Andre skalldyr, bløtdyr og pigghuder            4
Haifisk                                         4
Sjøpattedyr                                     1
Annen pelagisk fisk                             1
Name: count, dtype: int64 Art - gruppe
Torsk                                       57885810.0
Sei                                        

In [254]:
# Ser etter hvilke fisk som blir fisket med andre liner:
condition = df[df['Redskap FDIR'] == 'Andre liner']
species_weight_sum = condition.groupby('Art - gruppe')['Rundvekt'].sum().sort_values(ascending=False)
print(condition['Art - gruppe'].value_counts(), species_weight_sum)

Art - gruppe
Annen torskefisk                            11202
Torsk                                        7372
Steinbiter                                   6712
Hyse                                         6641
Uer                                          3236
Sei                                          2543
Annen flatfisk, bunnfisk og dypvannsfisk     2315
Blåkveite                                    2105
Skater og annen bruskfisk                    1786
Haifisk                                        58
Kongekrabbe, annen                              4
Brunalger                                       1
Name: count, dtype: int64 Art - gruppe
Torsk                                       33155234.0
Hyse                                        24829551.0
Annen torskefisk                            21104924.0
Steinbiter                                   4769635.0
Blåkveite                                    3210210.0
Sei                                           914214.0
Skater og annen br

In [255]:
# Ser etter hvilke fisk som blir fisket med reketrål:
condition = df[df['Redskap FDIR'] == 'Reketrål']
species_weight_sum = condition.groupby('Art - gruppe')['Rundvekt'].sum().sort_values(ascending=False)
print(condition['Art - gruppe'].value_counts(), species_weight_sum)

Art - gruppe
Dypvannsreke                                5838
Annen flatfisk, bunnfisk og dypvannsfisk    3968
Annen torskefisk                            3297
Torsk                                       2950
Sei                                         2769
Skater og annen bruskfisk                   1227
Hyse                                         497
Øyepål                                       461
Andre skalldyr, bløtdyr og pigghuder         338
Annen pelagisk fisk                          190
Uer                                          182
Steinbiter                                   177
Blåkveite                                    127
Kolmule                                      119
Vassild og strømsild                         112
Sild, annen                                   29
Makrell                                       26
Haifisk                                       19
Tobis og annen sil                             8
Mesopelagisk fisk                              2
Brunalg

In [256]:
# Ser etter hvilke fisk som blir fisket med snurpenot, ringnot et fartøy:
condition = df[df['Redskap FDIR'] == 'Snurpenot/ringnot']
species_weight_sum = condition.groupby('Art - gruppe')['Rundvekt'].sum().sort_values(ascending=False)
print(condition['Hovedart FAO'].value_counts(), species_weight_sum)

Hovedart FAO
Sild                    5539
Sei                     1279
Lodde                   1250
Makrell                  952
Torsk                    735
Brisling                 171
Tobis og annen sil       144
Hestmakrell               98
Kolmule                   64
Strømsild/Vassild         23
Strømsild                 20
Hyse                       7
Vassild                    5
Blåkveite                  5
Makrellstørje              2
Nordlig lysprikkfisk       1
Name: count, dtype: int64 Art - gruppe
Sild, annen                                 364481891.0
Lodde                                       177152131.0
Makrell                                     160745614.0
Sei                                          34768481.0
Kolmule                                      24500480.0
Tobis og annen sil                            5694001.0
Torsk                                         4394568.0
Kystbrisling                                  2688800.0
Annen pelagisk fisk                 

Etter å ha studert de ulike redskapene og hva som blir fisket med de aktuelle, ser vi også konturene av hvordan de blir brukt. Bunntrål er den mest ukritiske fangstmetoden og den tar med seg et større mangfold av fisk enn for eksempel snurrevad som kanskje er naturlig å sammenligne med. Man skulle tro av navnet at en reketrål var mer målrettet i fangsten, og at andelen reker var større i denne sammensetningen, men her kan vi også anta, at siden reker står ganske lavt i næringskjeden, så vil det være naturlig at det er en stor andel av "rovfisker" i miksen. Vi har ikke sett her hvor stor andel av vekten de andre fiskene utgjør

Redskaper:
    Snurpenot/ringnot:
        Dette er det redskapet som det blir fisket desidert mest fisk med, selv om det er mye flere registreringer av andre redskaper. Forklaring her kan være at at med Snurpenot/ringnot blir det fisket mest stimfisk som sild, lodde, makrell osv, som opptrer i stim og dermed er naturlig at man fanger mye av når man først har funnet stimen. Det trenger ikke nødvendigvis bety at det er et mer effektivt fiske.
    Bunntrål:
        Med dette redskapet blir det fisket mest torsk og sei, men også ganske mye av andre arter. Det er det redskapet som vi har undersøkt som har jevnest fordeling arter, som betyr at det er det redskapet som bidrar med mest bifangst.
    Snurrevad:
        Dette redskapet ligner på bunntrål, men det virker som det drives litt mer målrettet fiske med dette. Spredningen av fangst av forskjellige arter er ikke så stor.
    Andre liner:
    Reketrål:
    
Basert på denne infoen er det interessant å dykke videre inn i et bestemt redskap og art og muligens et bestemt område. 


In [257]:
print(df['Hovedområde start'].value_counts())

Hovedområde start
Eigersundbanken                  45349
Røstbanken til Malangsgrunnen    29099
Vest-Finnmark                    27711
Øst-Finnmark                     26995
Bjørnøya                         25265
Shetland                         24813
Nordkappbanken                   18635
Vikingbanken                     16600
Skagerrak                        16097
Storegga-Frøyabanken             14884
Helgelandsbanken                  8591
Sentralbanken                     5584
Storbanken                        4823
Thor Iversens Bank                4551
Vestfjorden (Lofoten)             4469
Sentrale Nordsjø                  3387
Hopen                             3161
Vest av Skottland (Hebridene)     2008
Sørlige Norskehav                 1990
Østlige Norskehav                 1936
Admiralityfeltet                  1909
Skolpenbanken                     1904
Gåsebanken                        1699
Storfjord/Hinlopenstredet         1363
Vest-Spitsbergen                  1242
Vest av

Lokasjon:
Det er ingen hemmelighet at fiske etter ulike arter foregår på ulike steder til ulikt tidspunkt, avhengig av hva man ønsker å fiske, og hvor mye. For eksempel har skreifiske i lofoten pågått i tusener av år. Storsilden har også tradisjonelt sett kommet inn til vestlandskysten på vinteren.
https://open-data-fiskeridirektoratet-fiskeridir.hub.arcgis.com/datasets/1b8c3425aaf643e9b2abd3bc70111697/explore

Sei:
Vi har valgt å fokusere på seifiske: I Norge er det hovedsaklig fiske etter to grupper av sei. En gruppe som beiter i havområdene mellom Norge og Skottland https://www.hi.no/hi/temasider/arter/sei-i-nordsjoen og en annen som beiter i kystområdene fra vestlandet til finnmark. https://www.hi.no/hi/temasider/arter/nordaustarktisk-sei.
De er rovfisker og vandrefisk og kan følge sild langt ut i norskehavet. 

Om vi skal fokusere på den nordøstarktiske seien, kan det se ut fra denne informasjon at det kan være lurt å fokusere på hovedområdene fra 0-7 minus 2, samt 28. Om vi skal inkludere Nordsjøseien, så kan vi inkludere områdene 40-43, men da beveger vi oss også litt utenfor de de norske områdene.

Videre er det 



In [258]:
# Dropper kolonner: Meldingstids
dropped_df = df.drop(columns=[
    #'Melding ID', Trengs muligens for å skille ulike tokter??
    'Meldingstidspunkt',  # Trenger ikke tidspunkt for melding når vi undersøker selve fisket. 
    'Meldingsdato',
    'Meldingsklokkeslett',
    'Starttidspunkt', 
    'Startdato', 
    'Startklokkeslett',
    'Startposisjon bredde', 
    'Startposisjon lengde',
    #'Hovedområde start (kode)', 
    'Hovedområde start',
    'Lokasjon start (kode)',
    'Havdybde start',
    #'Stopptidspunkt', # Det kan være interessant å se om det er sammenheng i tidspunkt på døgn og fangst.
    'Stoppdato', # All tidspunkt informasjon får vi i kolonnen over.
    'Stoppklokkeslett',
    'Varighet',
    'Fangstår', # Irrelevant når så og si all data er fra 2018
    #'Stopposisjon bredde', 
    #'Stopposisjon lengde',
    'Hovedområde stopp (kode)', 
    'Hovedområde stopp',
    'Lokasjon stopp (kode)',
    #'Havdybde stopp', 
    #'Trekkavstand',
    'Redskap FAO (kode)',
    'Redskap FAO',
    'Redskap FDIR (kode)',
    #'Redskap FDIR', 
    #'Hovedart FAO (kode)', 
    #'Hovedart FAO',
    'Hovedart - FDIR (kode)',
    #'Art FAO (kode)', 
    #'Art FAO',
    'Art - FDIR (kode)',
    'Art - FDIR',
    'Art - gruppe (kode)',
    #'Art - gruppe', # Kan være nyttig for å få oversikt
    #'Rundvekt', 
    'Lengdegruppe (kode)', # Vi har valgt å bruke Bruttotonnasje som størrelse/kapasitetmål på båtene. 
    'Lengdegruppe',
    #'Bruttotonnasje kombinert'
    'Bruttotonnasje 1969',
    'Bruttotonnasje annen',
    'Bredde',
    'Fartøylengde'
])
dropped_df.head()

Melding ID  Hovedområde start (kode)    Stopptidspunkt  \
0     1497177                       NaN        01.01.2018   
1     1497178                      20.0  31.12.2017 04:16   
2     1497178                      20.0  31.12.2017 04:16   
3     1497178                      20.0  31.12.2017 04:16   
4     1497178                      20.0  31.12.2017 04:16   

   Stopposisjon bredde  Stopposisjon lengde  Havdybde stopp  Trekkavstand  \
0              -60.400              -46.150            -805        5649.0   
1               74.914               15.969            -334        3970.0   
2               74.914               15.969            -334        3970.0   
3               74.914               15.969            -334        3970.0   
4               74.914               15.969            -334        3970.0   

  Redskap FDIR Hovedart FAO (kode)      Hovedart FAO Art FAO (kode)  \
0    Flytetrål                 KRI  Antarktisk krill            KRI   
1     Bunntrål                 HAD              Hyse            HAD   
2     Bunntrål                 HAD              Hyse            COD   
3     Bunntrål                 HAD              Hyse            GHL   
4     Bunntrål                 HAD              Hyse            POK   

            Art FAO      Art - gruppe  Rundvekt  
0  Antarktisk krill  Antarktisk krill  706714.0  
1              Hyse              Hyse    9594.0  
2             Torsk             Torsk    8510.0  
3         Blåkveite         Blåkveite     196.0  
4               Sei               Sei     134.0

In [274]:
# Lager en betingelse med aktuelle områder.
northeastern = [0, 1, 3, 4, 5, 6, 7, 28]
northeastern_ocean = [30, 34, 37, 39, 12, 36] # For å undersøke om det er noe særlig seifiske utenfor de mer kystnære strøk
northsea = [40, 41, 42, 43] # For å undersøke seifiske
areas_condition = dropped_df['Hovedområde start (kode)'].isin(northeastern)
# Lager en betingelse med aktuelle fiskearter
fish_types_conditions = dropped_df['Hovedart FAO'] == 'Sei'
# Lager en betingelse med aktuelle redskaper:
fish_tools_conditions = dropped_df['Redskap FDIR'] == 'Bunntrål'
conditions = areas_condition & fish_types_conditions & fish_tools_conditions
filtered_df = dropped_df[conditions]
filtered_df.describe()  # Får her et inntrykk av ekstremverdier og gjennomsnitt.

Melding ID  Hovedområde start (kode)  Stopposisjon bredde  \
count  2.703000e+04              27030.000000         27030.000000   
mean   1.646568e+06                  8.622494            66.726095   
std    9.436028e+04                  8.318941             4.139766   
min    1.497341e+06                  0.000000            59.647000   
25%    1.549231e+06                  4.000000            62.836000   
50%    1.665665e+06                  5.000000            68.030000   
75%    1.729002e+06                  7.000000            70.882000   
max    1.800286e+06                 28.000000            72.060000   

       Stopposisjon lengde  Havdybde stopp   Trekkavstand      Rundvekt  
count         27030.000000    27030.000000   27030.000000  27030.000000  
mean             13.021517     -192.309989   13474.779948   2676.027340  
std               8.355615       73.043725   11594.323343   5357.860741  
min               1.654000    -1276.000000       0.000000      0.000000  
25%               5.339000     -227.000000    4990.000000     77.000000  
50%              10.942000     -183.000000   10145.000000    260.000000  
75%              20.581000     -145.000000   19322.000000   2377.500000  
max              32.226000      -49.000000  233749.000000  86506.000000

Datacleaning:
Nullverdier.

Av df.describe leser vi først max og min verdier for å se om det er noen ekstreme eller usannsynlige verdier.
Rundvekt, 0 - registreringer. Ser at det er registreringer med verdien 0 i trekkavstand og rundvektkolonnen. Vi anser dette som feilregistreringer. Av trålens natur, at den soper store havområder, ser vi det som lite trolig at rundvekt 0 skal forekomme. Videre kan vi anta at registreringer med 0 i trekkavstand også er lite sannsynlig. 

Trekkavstand:
Vi ser at det ligger registreringer på over 200000 meter, det vil si 200 km eller 20 mil. Dette er store avstander for en trål som vanligvis opererer i snittfart på mellom 3,4 (6,3km/h) og 5 knop (9,26 km/h). I en snittfart på 10km/h så vil da trålen bruke over 20 timer. Med så lang varighet vil det være svært ugunstige forhold for fisken. 
Vi ser også at 5% av registreringene er over 35446m. Her ser vi også at med relativt høy fart, vil aktiviteten ha varighet på nærmere 4 timer. Ved å kutte registreringer over 50000 gir det oss en god margin på mulige feilregistreringer.

Dybde:
Vi ser det er registrert positive havdybde-verdier, i tillegg er det også registrert aktiviteter på -1276 dybde. Men de aller fleste aktivitetene er registrert mellom -227 og -145 meter. Fiskeridirektoratet beskriver at "Sei førekjem både pelagisk og som botnfisk, frå 0 til 300 meters djup." (https://www.hi.no/hi/temasider/arter/nordaustarktisk-sei) Vi kan anta at dette er enten en feilregistrering eller at fangsten er gjort når trålen er på vei opp av dypet. 



In [268]:
trekkavstand = filtered_df['Trekkavstand'] == 0
filtered_df[trekkavstand]

Melding ID  Hovedområde start (kode)    Stopptidspunkt  \
289        1497435                       5.0  01.01.2018 16:47   
290        1497435                       5.0  01.01.2018 16:47   
291        1497435                       5.0  01.01.2018 16:47   
292        1497435                       5.0  01.01.2018 16:47   
13516      1508015                       5.0  19.01.2018 16:47   
13517      1508015                       5.0  19.01.2018 16:47   
16358      1510200                       5.0  22.01.2018 19:04   
16359      1510200                       5.0  22.01.2018 19:04   
16360      1510200                       5.0  22.01.2018 19:04   
17349      1510864                       5.0  23.01.2018 17:44   
17350      1510864                       5.0  23.01.2018 17:44   
18450      1511569                       5.0  24.01.2018 17:03   
18451      1511569                       5.0  24.01.2018 17:03   
18452      1511569                       5.0  24.01.2018 21:18   
18453      1511569                       5.0  24.01.2018 21:18   
21754      1513721                       5.0  28.01.2018 02:51   
21755      1513721                       5.0  28.01.2018 02:51   
37931      1533312                       7.0  13.02.2018 01:53   
37932      1533312                       7.0  13.02.2018 01:53   
37933      1533312                       7.0  13.02.2018 01:53   
37934      1533312                       7.0  13.02.2018 01:53   
37935      1533312                       7.0  13.02.2018 01:53   
37936      1533312                       7.0  13.02.2018 01:53   
41165      1535772                      28.0  16.02.2018 17:52   
42632      1536811                      28.0  17.02.2018 00:30   
42633      1536811                      28.0  17.02.2018 00:30   
55247      1547353                      28.0  02.03.2018 13:41   
160383     1680201                       4.0  22.06.2018 04:29   
160384     1680201                       4.0  22.06.2018 04:29   
160385     1680201                       4.0  22.06.2018 04:29   
160386     1680201                       4.0  22.06.2018 04:29   
222213     1730383                       5.0  06.09.2018 14:42   
222439     1730467                       5.0  06.09.2018 04:24   
222440     1730467                       5.0  06.09.2018 04:24   
222441     1730467                       5.0  06.09.2018 10:31   
222442     1730467                       5.0  06.09.2018 10:31   
276480     1777445                       4.0  15.11.2018 09:36   
276481     1777445                       4.0  15.11.2018 09:36   
276482     1777445                       4.0  15.11.2018 09:36   
288763     1787521                       3.0  30.11.2018 02:45   
288764     1787521                       3.0  30.11.2018 02:45   
288765     1787521                       3.0  30.11.2018 02:45   
288766     1787521                       3.0  30.11.2018 02:45   
288767     1787521                       3.0  30.11.2018 02:45   
304640     1799900                       7.0  29.12.2018 09:33   
304641     1799900                       7.0  29.12.2018 09:33   
304642     1799900                       7.0  29.12.2018 09:33   

        Stopposisjon bredde  Stopposisjon lengde  Havdybde stopp  \
289                  68.491               11.567            -425   
290                  68.491               11.567            -425   
291                  68.491               11.567            -425   
292                  68.491               11.567            -425   
13516                69.450               15.800            -751   
13517                69.450               15.800            -751   
16358                69.500               15.833            -235   
16359                69.500               15.833            -235   
16360                69.500               15.833            -235   
17349                69.417               15.867            -178   
17350                69.417               15.867            -178   
18450                69

Ser av trekkavstand at det i 0-registreringene av trekkavstand også fanges fisk. Dette gir ikke mening med tanke på at det følgelig må trekkes en viss avstand for å få fisk.

In [271]:
rundvekt = filtered_df['Rundvekt'] == 0
filtered_df[rundvekt].head()

Melding ID  Hovedområde start (kode)    Stopptidspunkt  \
7538      1503695                       5.0  11.01.2018 08:35   
9677      1505102                       4.0  14.01.2018 20:42   
10271     1505399                       5.0  15.01.2018 19:16   
27484     1519324                      28.0  02.02.2018 11:52   
27485     1519324                      28.0  02.02.2018 11:52   

       Stopposisjon bredde  Stopposisjon lengde  Havdybde stopp  Trekkavstand  \
7538                70.694               17.278            -282       29217.0   
9677                70.569               17.249            -276        5359.0   
10271               70.573               17.227            -275       26163.0   
27484               61.137                2.724            -289       26840.0   
27485               61.137                2.724            -289       26840.0   

      Redskap FDIR Hovedart FAO (kode) Hovedart FAO Art FAO (kode)  \
7538      Bunntrål                 POK          Sei            GHL   
9677      Bunntrål                 POK          Sei            MZZ   
10271     Bunntrål                 POK          Sei            MZZ   
27484     Bunntrål                 POK          Sei            HAD   
27485     Bunntrål                 POK          Sei            HKE   

                Art FAO                              Art - gruppe  Rundvekt  
7538          Blåkveite                                 Blåkveite       0.0  
9677   Annen marin fisk  Annen flatfisk, bunnfisk og dypvannsfisk       0.0  
10271  Annen marin fisk  Annen flatfisk, bunnfisk og dypvannsfisk       0.0  
27484              Hyse                                      Hyse       0.0  
27485            Lysing                          Annen torskefisk       0.0

Ser også her at det er registrert forskjellige arter i 0-registrerningene på rundvekt. Det er da fornuftig å anta at dette enten er feilregistreringer eller fangst uten betydning.

Ser av value_counts at de mest vanlige redskapene er bunntrål og snurrevad. Her er det så mye data at vi fokuserer videre på dette. (Usikker på å inkludere snurrevad.

df['Redskap

In [ ]:
# Slår sammen Bruttotonnasje 1969 og bruttotonnasje annen:
df.loc[:, 'Bruttotonnasje kombinert'] = df['Bruttotonnasje 1969'].fillna(df['Bruttotonnasje annen'])

In [ ]:
# Kjører df.describe()
df.describe()

In [ ]:
df.drop(columns=[#'Melding ID', Trengs muligens for å skille ulike tokter??
                 'Meldingstidspunkt', 
                 'Meldingsdato',
                 'Meldingsklokkeslett', 
                 'Starttidspunkt', 
                 'Startdato',
                 'Startklokkeslett', 
                 #'Startposisjon bredde', 
                 #'Startposisjon lengde',
                 #'Hovedområde start (kode)', 
                 'Hovedområde start',
                 'Lokasjon start (kode)', 
                 'Havdybde start', 
                 #'Stopptidspunkt', Det kan være interessant å se om det er sammenheng i tidspunkt på døgn og fangst.
                 'Stoppdato', 
                 'Stoppklokkeslett', 
                 'Varighet', 
                 'Fangstår',
                 'Stopposisjon bredde', 
                 'Stopposisjon lengde',
                 'Hovedområde stopp (kode)', 
                 'Hovedområde stopp',
                 'Lokasjon stopp (kode)', 
                 #'Havdybde stopp', 
                 #'Trekkavstand',
                 'Redskap FAO (kode)', 
                 'Redskap FAO', 
                 'Redskap FDIR (kode)',
                 #'Redskap FDIR', 
                 #'Hovedart FAO (kode)', 
                 #'Hovedart FAO',
                 'Hovedart - FDIR (kode)', 
                 #'Art FAO (kode)', 
                 #'Art FAO',
                 'Art - FDIR (kode)', 
                 'Art - FDIR', 
                 'Art - gruppe (kode)',
                 #'Art - gruppe', 
                 #'Rundvekt', 
                 #'Lengdegruppe (kode)', 
                 #'Lengdegruppe',
                 #'Bruttotonnasje kombinert'
                 'Bruttotonnasje 1969', 
                 'Bruttotonnasje annen', 
                 #'Bredde',
                 #'Fartøylengde'
                 ])
df

In [ ]:
# Antall registreringer med dei forskjellige Hovedarter.
print(df['Hovedart FAO'].value_counts())

In [ ]:
# Få oversikt over hovedområder.
hovedområder = df.groupby(df['Hovedområde start'], as_index= False).agg(
    {'Hovedområde start': 'first', 'Hovedområde start (kode)': 'first'}
)
print(hovedområder)

In [ ]:
# Undersøker hvor det er flest aktiviteter
print(df['Hovedområde start'].value_counts())

In [ ]:
# Undersøker hvilke arter som er de mest frekvente i Eigersundbanken.
eigersundbanken = df[df['Hovedområde start'] == 'Eigersundbanken']
eigersundbanken['Hovedart FAO'].value_counts()

In [ ]:
# Antall registreringer av fisk som ikkje er hovedart.
condition = df[(df['Hovedart FAO'] != df['Art FAO'])]
condition['Art FAO'].value_counts()

In [ ]:
# Aktuelle redskaper å bruke: Bunntrål, Snurrevad og muligens Andre Liner. evt, dobbeltål
# Unngå garn.
# Aktuelle fisketyper å bruke hovedfangst: Sei
fish_types = ['Sei']
# Aktuelle Redskaper: Bunntrål, Snurrevad, Andre liner, (Dobbeltrål)
tools = ['Bunntrål']  # , 'Snurrevad', 'Andre liner'
# Aktuelle områder:
#areas = [x for x in range(0, 70)] # alle hovedområder minus antarktis og fiske i andre hav.
areas = [8, 42, 28, 7]

condition_3 = df['Hovedområde start (kode)'].isin(areas)
condition_1 = df['Hovedart FAO'].isin(fish_types) 
condition_2 = df['Redskap FDIR'].isin(tools)
combined_condition = condition_1 & condition_2 & condition_3
filtered_df = df[combined_condition].copy()
filtered_df.info()

In [ ]:
filtered_df.loc[:, 'Bruttotonnasje kombinert'] = filtered_df['Bruttotonnasje 1969'].fillna(filtered_df['Bruttotonnasje annen'])

for start_stop in ['Startklokkeslett', 'Stoppklokkeslett']:
    # Konvertere klokkeslett til relevant verdi
    filtered_df[start_stop] = pd.to_datetime(filtered_df[start_stop], format='%H:%M').dt.strftime('%H-%M')
    # Konverter til Desimal-timer.
    filtered_df[f'{start_stop}_DecimalHours'] = filtered_df[start_stop].str.split('-').apply(lambda x: int(x[0]) + int(x[1]) / 60)
    # Konverter til Radianer.
    filtered_df[f'{start_stop}_Radians'] = filtered_df[f'{start_stop}_DecimalHours'] * (2 * np.pi / 24)
    # Regne ut sinus og cosinus.
    filtered_df[f'{start_stop}_TimeSin'] = np.sin(filtered_df[f'{start_stop}_Radians'])
    filtered_df[f'{start_stop}_TimeCos'] = np.cos(filtered_df[f'{start_stop}_Radians'])

In [ ]:
filtered_df = filtered_df.drop(columns = [
    'Fangstår', 
    #'Melding ID', 
    'Meldingstidspunkt', 
    'Meldingsdato',
    'Meldingsklokkeslett', 
    'Starttidspunkt',
    'Startklokkeslett',
    #'Hovedområde start (kode)', 
    'Hovedområde start',
    'Lokasjon start (kode)', 
    'Havdybde start', 
    'Stopptidspunkt',
    'Varighet',
    'Stoppdato', 
    'Startposisjon bredde',
    'Startposisjon bredde',
    #'Stoppklokkeslett', 
    #'Stopposisjon bredde', 
    #'Stopposisjon lengde',
    'Hovedområde stopp (kode)', 
    'Hovedområde stopp',
    'Lokasjon stopp (kode)', 
    'Redskap FAO (kode)', 
    'Redskap FDIR (kode)',
    'Hovedart FAO (kode)',
    'Hovedart - FDIR (kode)', 
    'Art FAO (kode)', 
    #'Art FAO',
    'Art - FDIR (kode)', 
    'Art - FDIR', 
    'Art - gruppe (kode)',
    #'Art - gruppe', 
    'Lengdegruppe (kode)', 
    'Lengdegruppe',
    'Bredde',
    'Fartøylengde',
    'Bruttotonnasje 1969',
    'Bruttotonnasje annen'
    ])

In [ ]:
def join_unique_values(series):
    return ', '.join(series.dropna().astype(str).unique())

# Gruppere df og slå sammen rader som kommer av samme aktivitet:
grouped_df = filtered_df.groupby(['Melding ID', 'Stoppklokkeslett'], as_index= False).agg({
    'Stopposisjon lengde': 'first',
    'Stopposisjon bredde': 'first',
    'Stoppklokkeslett': 'first',
    'Stoppklokkeslett_DecimalHours': 'first',
    'Stoppklokkeslett_Radians': 'first',
    'Stoppklokkeslett_TimeSin': 'first',
    'Hovedområde start (kode)': 'first',
    'Havdybde stopp': 'last',
    'Trekkavstand': 'first',  
    'Redskap FDIR': join_unique_values,
    'Hovedart FAO': 'first',  
    'Art FAO': join_unique_values,
    'Art - gruppe': join_unique_values,
    'Rundvekt': 'sum',   
    'Bruttotonnasje kombinert': 'first'
})
grouped_df

In [ ]:
# Droppe tomme rader.
cleaned_df = grouped_df.copy().dropna()
# Konvertere positive Havdybdeverdier
cleaned_df['Havdybde stopp'] = cleaned_df['Havdybde stopp'].abs()
# Ta vekk havdybder på over 10 m, for å fjerne usannsynlige tråledybder.
cleaned_df = cleaned_df[cleaned_df['Havdybde stopp'] > 10]
# Droppe trekkavstand på over ei viss lengd.
cleaned_df = cleaned_df[(cleaned_df['Trekkavstand'] > 10) & (cleaned_df['Trekkavstand'] < 60000)]

# Droppe ekstreme Bruttotonnasjeverdier. Under
cleaned_df = cleaned_df[(cleaned_df['Bruttotonnasje kombinert'] > 400) & (cleaned_df['Bruttotonnasje kombinert'] < 4000)]
# Droppe registreringer i områder som ikkje er i Norsk farvatn. Startposisjon med mindre enn 0.
cleaned_df = cleaned_df[cleaned_df['Stopposisjon lengde'] > 0]
cleaned_df = cleaned_df[cleaned_df['Stopposisjon bredde'] > 0]
cleaned_df.info()

In [ ]:
# Lager en liste med features
all_features = [
    'Stopposisjon lengde', 
    'Stopposisjon bredde', 
    'Stoppklokkeslett_Radians', 
    'Havdybde stopp', 
    'Trekkavstand',  
    'Rundvekt',
    'Bruttotonnasje kombinert'
]

cleaned_df[all_features]

In [ ]:
# Importerer verktøy for å kalkulere feil i modellen - gjennomsnittlig feil?
from sklearn.metrics import mean_absolute_error

# Lager en for-loop for å generere resultater for ulike target-values.
for predict in all_features:
    features = [f for f in all_features if f != predict]
    X = cleaned_df[features]
    y = cleaned_df[predict]

    # Definerer X og y, X er features, y er målverdien (target value)
    X = cleaned_df[features]
    y = cleaned_df[predict]

    # Splitter settet i trainsett og testsett - Da kan man sammenligne prediksjonene mot et test-sett
    X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=1)

    # Velger MLmodell
    fishery_model = DecisionTreeRegressor(random_state=1)

    # Trener(?) modellen på train-settet
    fishery_model.fit(X_train, y_train)

    # Prediksjoner - predikerer tar X-val verdiene og predikerer rundvekt basert på disse tallene.
    val_predictions = fishery_model.predict(X_val)

    # Sammenligner et utvalg (X-val.head()) av prediksjoner mot relaterte y_val.head()
    print(f'--------- Target value: {predict} -------')
    print(fishery_model.predict(X_val.head()))
    print(y_val.head())

    # Tar inn y_val verdiene og regner ut gjennomsnittlig feil mellom prediksjonene og test-verdiene(val_y)
    val_mae = mean_absolute_error(y_val, val_predictions)
    print(f'\nMean absolute error: {val_mae}')
    print(f'\nScore: {fishery_model.score(X_val, y_val)}\n')


In [ ]:
for predict in all_features:
    features = [f for f in all_features if f != predict]
    X = cleaned_df[features]
    y = cleaned_df[predict]

    # Definerer X og y, X er features, y er målverdien (target value)
    X = cleaned_df[features]
    y = cleaned_df[predict]

    # Splitter settet i trainsett og testsett - Da kan man sammenligne prediksjonene mot et test-sett
    X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=1)

    # Velger MLmodell
    fishery_model = linear_model.LinearRegression()

    # Trener(?) modellen på train-settet
    fishery_model.fit(X_train, y_train)

    # Prediksjoner - predikerer tar X-val verdiene og predikerer rundvekt basert på disse tallene.
    val_predictions = fishery_model.predict(X_val)

    # Sammenligner et utvalg (X-val.head()) av prediksjoner mot relaterte y_val.head()
    print(f'--------- Target value: {predict} -------')
    print(fishery_model.predict(X_val.head()))
    print(y_val.head())

    # Tar inn y_val verdiene og regner ut gjennomsnittlig feil mellom prediksjonene og test-verdiene(val_y)
    val_mae = mean_absolute_error(y_val, val_predictions)
    print(f'\nMean absolute error: {val_mae}')
    print(f'\nScore: {fishery_model.score(X_val, y_val)}\n')